Replace WebTransportDatagramDuplexStream.writable with createWritable

achristensen07 opened the pull request:
#### 2b5ed40d5731db32441ac0738d69bf2b3765122b
<pre>
Replace WebTransportDatagramDuplexStream.writable with createWritable
<a href="https://bugs.webkit.org/show_bug.cgi?id=302318">https://bugs.webkit.org/show_bug.cgi?id=302318</a>
<a href="https://rdar.apple.com/164461818">rdar://164461818</a>

Reviewed by Abrar Rahman Protyasha, Brent Fulgham, and Matthew Finkel.

This matches the current spec.  Other browsers haven&apos;t done this yet,
but I&apos;m told they intend to.  This introduces WebTransportSendGroup.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp:
(WebCore::DatagramSink::DatagramSink):
(WebCore::DatagramSink::attachTo):
(WebCore::DatagramSink::write):
* Source/WebCore/Modules/webtransport/DatagramSink.h:
(WebCore::DatagramSink::create):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::create):
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::datagramsWritableCreated):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
(WebCore::WebTransport::createSendGroup):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp:
(WebCore::WebTransportDatagramDuplexStream::create):
(WebCore::WebTransportDatagramDuplexStream::WebTransportDatagramDuplexStream):
(WebCore::WebTransportDatagramDuplexStream::attachTo):
(WebCore::WebTransportDatagramDuplexStream::createWritable):
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h:
(WebCore::WebTransportDatagramDuplexStream::writable): Deleted.
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.cpp: Added.
(WebCore::WebTransportDatagramsWritable::create):
(WebCore::WebTransportDatagramsWritable::WebTransportDatagramsWritable):
(WebCore::WebTransportDatagramsWritable::sendGroup):
(WebCore::WebTransportDatagramsWritable::setSendGroup):
(WebCore::WebTransportDatagramsWritable::sendOrder):
(WebCore::WebTransportDatagramsWritable::setSendOrder):
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.h: Copied from Source/WebCore/Modules/webtransport/DatagramSink.h.
(isType):
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportSendGroup::create):
(WebCore::WebTransportSendGroup::WebTransportSendGroup):
(WebCore::WebTransportSendGroup::getStats):
(WebCore::WebTransportSendGroup::transport):
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.h: Copied from Source/WebCore/Modules/webtransport/DatagramSink.h.
(WebCore::WebTransportSendGroup::identifier const):
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendOptions.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendOptions.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::create):
(WebCore::WebTransportSendStream::WebTransportSendStream):
(WebCore::WebTransportSendStream::getStats):
(WebCore::WebTransportSendStream::sendGroup):
(WebCore::WebTransportSendStream::setSendGroup):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::sendDatagram):
(WebCore::WorkerWebTransportSession::getSendGroupStats):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/Modules/webxr/WebXRFrame.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::getSendGroupStats):
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::getSendGroupStats):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Datagram)):
(TestWebKitAPI::TEST(WebTransport, NetworkProcessCrash)):
(TestWebKitAPI::TEST(WebTransport, CreateStreamsBeforeReady)):

Canonical link: <a href="https://commits.webkit.org/303236@main">https://commits.webkit.org/303236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6301a4bdfd91d893a77dd563e60f0bb3bd2c99cd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131789 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139299 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133659 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4211 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4043 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134735 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118037 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81517 "Passed tests") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/131138 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82520 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36162 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141944 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/3950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36741 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4031 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109268 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3000 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114318 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20489 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4004 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67451 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3964 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/Modules/webtransport/DatagramSink.cpp:
```diff
@@ -29,19 +29,22 @@
 
-DatagramSink::DatagramSink() = default;
+DatagramSink::DatagramSink(RefPtr<WebTransportSession>&& session)
+    : m_session(WTFMove(session)) { }
```
Isn't `m_session` a `ThreadSafeWeakPtr`? Why do we move a `RefPtr` in and then make that assignment? Is that a rule for SaferCPP?

To me, it seems simpler to just pass a `WebTransportSession` reference (or pointer) if we're just going to bind a `WeakPtr` to it.

achristensen07 replied:
I should probably just make the parameter a ThreadSafeWeakPtr&&.

sysrqb commented on Source/WebCore/Modules/webxr/WebXRFrame.cpp:
```diff
@@ -103,11 +104,6 @@ bool WebXRFrame::mustPosesBeLimited(const WebXRSpace& space, const WebXRSpace& b
-    TransformationMatrix transform;
-    bool emulatedPosition { false };
-};
-
```
This doesn't seem intentional

achristensen07 replied:
This is necessary to fix the build with unified builds.

sysrqb commented on Source/WebCore/Modules/webxr/WebXRFrame.h:
```diff
@@ -85,7 +86,11 @@ class WebXRFrame : public RefCounted<WebXRFrame> {
+        TransformationMatrix transform;
+        bool emulatedPosition { false };
+    };
+
```
Or these

achristensen07 replied:
This is necessary to fix the build with unified builds.

aprotyas commented on Source/WebCore/Modules/webtransport/WebTransport.h:
```diff
@@ -134,6 +138,7 @@ class WebTransport : public WebTransportSessionClient, public ActiveDOMObject {
     const Ref<WebTransportReceiveStreamSource> m_receiveStreamSource;
     const Ref<WebTransportBidirectionalStreamSource> m_bidirectionalStreamSource;
     HashMap<WebTransportStreamIdentifier, Ref<WebTransportReceiveStreamSource>> m_readStreamSources;
+    ThreadSafeWeakHashSet<WebTransportDatagramsWritable> m_datagramsWritables;
```
It seems like you're not adding any writables to this? do you not somehow want to add to this set after `WebTransportDatagramDuplexStream::createWritable()`?

achristensen07 replied:
Indeed!

brentfulgham commented on Source/WebCore/Modules/webxr/WebXRFrame.h:
```diff
@@ -29,6 +29,7 @@
 
 #include "DOMHighResTimeStamp.h"
 #include "PlatformXR.h"
+#include "TransformationMatrix.h"
```
Seems like this could be a forward declaration?

achristensen07 replied:
There is a TransformationMatrix member variable.  This include is needed to fix compiling errors.

brentfulgham reviewed (approved):
This seems reasonable. Approving since you are the expert in this space, and these code changes look appropriate and clear. r=me.

aprotyas reviewed (approved):
I would maybe also add a test to verify cleanup when a WebTransport closes with active writer(s)?

sysrqb commented on Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.idl:
```diff
@@ -28,8 +28,8 @@
 ] interface WebTransportDatagramDuplexStream {
+    [CallWith=CurrentScriptExecutionContext] WebTransportDatagramsWritable createWritable(optional WebTransportSendOptions options = {});
     readonly attribute ReadableStream readable;
-    readonly attribute WritableStream writable;
```
Should we keep this for interop, or not worth it?

achristensen07 replied:
I specifically asked people to ask about this at IETF, and I was specifically told that other browsers intend to make this very change to match the spec, which intentionally removed it.

sysrqb reviewed (approved):
Can you also mention in the comment message that this patch also adds support for sendGroup?

achristensen07 committed:
Replace WebTransportDatagramDuplexStream.writable with createWritable
https://bugs.webkit.org/show_bug.cgi?id=302318
rdar://164461818

Reviewed by Abrar Rahman Protyasha, Brent Fulgham, and Matthew Finkel.

This matches the current spec.  Other browsers haven't done this yet,
but I'm told they intend to.  This introduces WebTransportSendGroup.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp:
(WebCore::DatagramSink::DatagramSink):
(WebCore::DatagramSink::attachTo):
(WebCore::DatagramSink::write):
* Source/WebCore/Modules/webtransport/DatagramSink.h:
(WebCore::DatagramSink::create):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::create):
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::datagramsWritableCreated):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
(WebCore::WebTransport::createSendGroup):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp:
(WebCore::WebTransportDatagramDuplexStream::create):
(WebCore::WebTransportDatagramDuplexStream::WebTransportDatagramDuplexStream):
(WebCore::WebTransportDatagramDuplexStream::attachTo):
(WebCore::WebTransportDatagramDuplexStream::createWritable):
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h:
(WebCore::WebTransportDatagramDuplexStream::writable): Deleted.
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.cpp: Added.
(WebCore::WebTransportDatagramsWritable::create):
(WebCore::WebTransportDatagramsWritable::WebTransportDatagramsWritable):
(WebCore::WebTransportDatagramsWritable::sendGroup):
(WebCore::WebTransportDatagramsWritable::setSendGroup):
(WebCore::WebTransportDatagramsWritable::sendOrder):
(WebCore::WebTransportDatagramsWritable::setSendOrder):
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.h: Copied from Source/WebCore/Modules/webtransport/DatagramSink.h.
(isType):
* Source/WebCore/Modules/webtransport/WebTransportDatagramsWritable.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportSendGroup::create):
(WebCore::WebTransportSendGroup::WebTransportSendGroup):
(WebCore::WebTransportSendGroup::getStats):
(WebCore::WebTransportSendGroup::transport):
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.h: Copied from Source/WebCore/Modules/webtransport/DatagramSink.h.
(WebCore::WebTransportSendGroup::identifier const):
* Source/WebCore/Modules/webtransport/WebTransportSendGroup.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendOptions.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendOptions.idl: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.idl.
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::create):
(WebCore::WebTransportSendStream::WebTransportSendStream):
(WebCore::WebTransportSendStream::getStats):
(WebCore::WebTransportSendStream::sendGroup):
(WebCore::WebTransportSendStream::setSendGroup):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::sendDatagram):
(WebCore::WorkerWebTransportSession::getSendGroupStats):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/Modules/webxr/WebXRFrame.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::getSendGroupStats):
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::getSendGroupStats):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Datagram)):
(TestWebKitAPI::TEST(WebTransport, NetworkProcessCrash)):
(TestWebKitAPI::TEST(WebTransport, CreateStreamsBeforeReady)):

Canonical link: https://commits.webkit.org/303236@main
